### PR TITLE
Add expire() in PoolBase

### PR DESCRIPTION
**Pull request by lgbo-ustc**
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add a new method `expire()` in PoolBase which is used to reallocate an invalid object in the pool.

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by tavplubix** [comment]
But what is the purpose of this change? New method in never used as for now.

**Comment by lgbo-ustc** [comment]
> But what is the purpose of this change? New method in never used as for now.

We use it in another developing feature. I want to  commit it alone

File: src/Common/PoolBase.h [hunk 1]
```diff
@@ -87,6 +88,15 @@ class PoolBase : private boost::noncopyable
         Object & operator*() &              { return *data->data.object; }
         const Object & operator*() const &  { return *data->data.object; }
 
+        /**
+         * Expire an object to make it reallocated later.
+         */
+        void expire()
+        {
+            std::unique_lock lock(data->data.pool.mutex);
+            data->data.is_expired = true;
```
Comment (kitaisreal) [inline src/Common/PoolBase.h:97]: Should this bool be atomic ?

File: src/Common/PoolBase.h [hunk 2]
```diff
@@ -113,7 +123,17 @@ class PoolBase : private boost::noncopyable
         {
             for (auto & item : items)
                 if (!item->in_use)
-                    return Entry(*item);
+                {
+                    if (!item->is_expired)
```
Comment (kitaisreal) [inline src/Common/PoolBase.h:127]: Consider to wrap `for` loop in braces, because inner part not 2 lines now.

Comment (kitaisreal) [inline src/Common/PoolBase.h:127]: Please add `likely(!item->is_expired)`.

File: src/Common/PoolBase.h [hunk 3]
```diff
@@ -139,6 +159,12 @@ class PoolBase : private boost::noncopyable
             items.emplace_back(std::make_shared<PooledObject>(allocObject(), *this));
     }
 
+    inline size_t aliveSize()
```
Comment (kitaisreal) [inline src/Common/PoolBase.h:162]: This returns just items size, consider to rename it into `size`.

**Review by kitaisreal — CHANGES_REQUESTED** [review]
Seems that pull request is ok, it does not make any significant changes to interfaces.

**Review by kitaisreal — APPROVED** [review]

**Pull request merged**